core/src/main/scala/org/apache/spark/deploy/security/README.md

```diff
@@ -0,0 +1,238 @@
+API) and, also, keeping them renewed while the app is running. These tokens are used not just
+by the application; they are also used by YARN itself to implement features like log collection
+and aggregation.
+
+But this has a few caveats.
+
+
+1. Who renews the tokens?
+
+This is handled mostly transparently by the Hadoop libraries in the case of YARN. Some services have
+the concept of a token "renewer". This "renewer" is the name of the principal that is allowed to
+renew the DT. When submitting to YARN, that will be the principal that the YARN service is running
+as, which means that the client application needs to know that information.
+
+For other resource managers, the renewer mostly does not matter, since there is no service that
+is doing the renewal. Except that it sometimes leaks into library code, such as in SPARK-20328.
+
+
+2. What tokens are renewed?
+
+This is probably the biggest caveat.
+
+As discussed in the previous section, DTs are service-specific, and require service-specific
+libraries for creation *and* renewal. This means that for YARN to be able to renew application
+tokens, YARN needs:
+
+* The client libraries for all the services the application is using
+* Information about how to connect to the services the application is using
+* Permissions to connect to those services
+
+In reality, though, most of the time YARN has access to a single HDFS cluster, and that will be
+the extent of its DT renewal features. Any other tokens sent to YARN will be distributed to
+containers, but will not be renewed.
+
+This means that those tokens will expire way before their max lifetime, unless some other code
+takes care of renewing them.
+
+And, to make matters worse, not all client libraries even implement token renewal. To use the
+example of a service supported by Spark, the `renew()` method of HBase tokens is a no-op. So
+the only way to "renew" an HBase token is to create a new one.
+
+
+3. What happens when tokens expire for good?
+
+The final caveat is that DTs have a maximum life, regardless of renewal. And after that deadline
+is met, you need to create new tokens to be able to connect to the services. That means you need
+the ability to connect to the service without a delegation token, which requires some form of
+authentication aside from DTs.
+
+This is especially important for long-running applications that run unsupervised. They must be
+able to keep on going without having someone logging into a terminal and typing a password every
+few days.
+
+
+## DT Renewal in Spark
+
+Because of the issues explained above, Spark implements a different way of doing renewal. Spark's
+solution is a compromise: it targets the lowest common denominator, which is services like HBase
+that do not support actual token renewal.
+
+In Spark, DT "renewal" is enabled by giving the application a Kerberos keytab. A keytab is
+basically your Kerberos password written into a plain text file, which is why it's so sensitive.
```

Comment:
I think we should explicitly state the security ramifications of this.  distributed a keytab means that if its compromised someone can get tokens for any service as you.

Reply:
Explain more than "it's like writing your password to a file"?

Reply:
Yes, I can see people not understanding that is a security concern.   If the file is written to HDFS with permissions restrictive, that might seem ok, but many security people would disagree.  Perhaps just add another sentence talking about this is a security concern because if the keytab is compromised people can get tokens for any service for an indefinite amount of time.  So users should evaluate if using this approach is ok.

Reply:
It sounds like you want to warn users of how dangerous this can be. This is not end-user documentation, so even if I make that statement here, it will probably not have the effect you want.